### PR TITLE
Studgradebook

### PR DESCRIPTION
**Pull request by dtracers**
A student gradebook that allows students to view their student grades

This depends on #436

### CODE REVIEW

File: CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+(function() {
+
+    var defaultCategory = 'All Assignments';
+
+    $(document).ready(function() {
+        CourseSketch.dataManager.waitForDatabase(function() {
+            var courseId = CourseSketch.dataManager.getState('gradebookCourseid');
```
Comment (DonaldMcRonald) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js]: shouldn't i in Id be capitalized??

File: CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    CourseSketch.gradeBook.loadGrades = function(courseId) {
+        CourseSketch.dataManager.getCourse(courseId, function(course) {
+            CourseSketch.gradeBook.course = course;
+            var categoryList = [ defaultCategory ];
+            var assignmentList = course.assignmentList;
+            CourseSketch.dataManager.getAssignments(assignmentList, undefined, function(assignments) {
+                // loads all of the grades
+                CourseSketch.dataManager.getAllAssignmentGrades(courseId, function(gradeList) {
+
```
Comment (DonaldMcRonald) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js]: delete this line

File: CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+     * This populates the map of grades for each student and, calls student adding function if student has not been added yet.
+     *
+     * @param {List<ProtoGrade>} listGrades The list of grades from the server.
+     * @param {Map<String, Element>} assignmentMap Map of assignmentIDs to table row.
+     * @param {HTMLTable} table The grade table on the webpage.
+     * @return {List<ProtoGrade>} grades that were not displayed. This is because the users do not exist anymore in the course roster.
+     */
+    CourseSketch.gradeBook.populateGrades = function(listGrades, assignmentMap, table) {
+        var gradesNotShown = [];
+        for (var i = 0; i < listGrades.length; i++) {
```
Comment (DonaldMcRonald) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js]: for each loop please

File: CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+     * Creates the assignment header while populating the assignment map.
+     *
+     * @param {List<String>} categoryList List of category names.
+     * @param {Map<String, Element>} categoryMap Map of category names to corresponding table element.
+     * @param {HTMLTable} table The grade table on the webpage.
+     */
+    function createCategoryHeader(categoryList, categoryMap, table) {
+        var template = document.querySelector('#expandableTemplate');
+        var container = document.importNode(template.content, true);
+        for (var i = 0; i < categoryList.length; i++) {
```
Comment (DonaldMcRonald) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js]: for each loop please

File: CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+     * @param {Map<String, Element>} categoryMap Map of category names to corresponding table element.
+     * @param {Element} table The grade table on the webpage.
+     */
+    function createAssignmentHeader(assignmentList, assignmentMap, categoryMap, table) {
+        var template = document.querySelector('#expandableTemplate');
+        var container = document.importNode(template.content, true);
+
+        var gradeTemplate = document.querySelector('#gradeHeader');
+        var gradeContainer = document.importNode(gradeTemplate.content, true);
+        for (var i = 0; i < assignmentList.length; i++) {
```
Comment (DonaldMcRonald) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js]: for each loop please

File: CourseSketchProjects/coursesketchwebclient/src/test/src/student/gradebook/studentGradebookFakePage.html [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+            var grade = CourseSketch.PROTOBUF_UTIL.ProtoGrade();
+            grade.courseId = courseId;
+            grade.userId = studentList[0];
+            grade.assignmentId = assignmentList[1];
+            grade.currentGrade = 80;
+            gradeList.push(grade);
+
+            for (var i = 0; i < studentList.length; i++) {
+                for (var j = 0; j < assignmentList.length; j++) {
```
Comment (DonaldMcRonald) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/student/gradebook/studentGradebookFakePage.html]: for each loop for both these loops please

File: CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js (hunk 1 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/student/gradebook/studentGradebook.js]: not right now

**Pull request merged**